WearOAuth/oauth-device-grant/src/main/java/com/example/android/wearable/oauth/devicegrant/AuthDeviceGrantViewModel.kt

```diff
@@ -0,0 +1,157 @@
+    // it possible to send an event to the Activity, which can then fire the Remote Intent.
+    private val fireRemoteIntentChannel = Channel<String>(Channel.BUFFERED)
+    val fireRemoteIntentFlow = fireRemoteIntentChannel.receiveAsFlow()
+
+    fun startAuthFlow() {
+        viewModelScope.launch(Dispatchers.IO) {
+            // Step 1: Retrieve the verification URI
+            showStatus("Starting authorization... Switch to your phone to authenticate.")
+            val verificationInfo = retrieveVerificationInfo().getOrElse {
+                showStatus("Authorization failed")
+                return@launch
+            }
+
+            // Step 2: Show the pairing code & open the verification URI on the paired device
+            showStatus("code: ${verificationInfo.userCode}")
+            fireRemoteIntentChannel.send(verificationInfo.verificationUri)
+
+            // Step 3: Poll the Auth server for the token
+            val token = retrieveToken(verificationInfo.deviceCode, verificationInfo.interval)
+
+            // Step 4: Use the token to make an authorized request
+            val userName = retrieveUserProfile(token).getOrElse {
+                showStatus("Authorization failed")
+                return@launch
+            }
+
+            showStatus("User profile retrieved. Welcome $userName!")
+        }
+    }
+
+    // The response data when retrieving the verification
+    data class VerificationInfo(
+        val verificationUri: String,
+        val userCode: String,
+        val deviceCode: String,
+        val interval: Int
+    )
+
+    /**
+     * Retrieve the information needed to verify the user. When performing this request, the server
+     * generates a user & device code pair. The user code is shown to the user and opened on the
+     * paired device. The device code is passed while polling the OAuth server.
+     */
+    private fun retrieveVerificationInfo(): Result<VerificationInfo> {
+        Log.d(TAG, "Retrieving verification info...")
+        return doPostRequest(
+            url = "https://oauth2.googleapis.com/device/code",
+            params = mapOf(
+                "client_id" to CLIENT_ID,
+                "scope" to "https://www.googleapis.com/auth/userinfo.profile"
+            )
+        ).map { responseJson ->
+            VerificationInfo(
+                verificationUri = responseJson.getString("verification_url"),
+                userCode = responseJson.getString("user_code"),
+                deviceCode = responseJson.getString("device_code"),
+                interval = responseJson.getInt("interval")
+            )
+        }
+    }
+
+    /**
+     * Poll the Auth server for the token. This will only return when the user has finished their
+     * authorization flow on the paired device.
+     *
+     * For this sample the various exceptions aren't handled.
+     */
+    private suspend fun retrieveToken(deviceCode: String, interval: Int): String {
```

Comment:
You can add `tailrec` here to avoid a large stack without manually converting the recursive call to an iterative one.